File: lumie_backend/app/services/profile_service.py
"""Profile service for user profile management."""
from datetime import datetime
from typing import Optional

from fastapi import HTTPException, status

from ..core.database import get_database
from ..models.user import (
    AccountRole,
    TeenProfileCreate,
    ParentProfileCreate,
    ProfileUpdate,
    UserProfile,
    HeightData,
    WeightData,
)


class ProfileService:
    """Service for managing user profiles."""

    async def create_teen_profile(self, user_id: str, data: TeenProfileCreate) -> UserProfile:
        """Create a teen profile."""
        db = get_database()

        # Get user and verify role
        user = await db.users.find_one({"user_id": user_id})
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found"
            )

        if user.get("role") != AccountRole.TEEN.value:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="User is not a teen account"
            )

        # Check if profile already exists
        existing_profile = await db.profiles.find_one({"user_id": user_id})
        if existing_profile:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Profile already exists. Use update endpoint."
            )

        # Validate age
        if data.age < 13:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Users must be 13 years or older to create an account"
            )

        now = datetime.utcnow()

        # Create profile document
        profile_doc = {
            "user_id": user_id,
            "role": AccountRole.TEEN.value,
            "name": data.name,
            "age": data.age,
            "height": {"value": data.height.value, "unit": data.height.unit.value},
            "weight": {"value": data.weight.value, "unit": data.weight.unit.value},
            "icd10_code": data.icd10_code,
            "advisor_name": data.advisor_name,
            "created_at": now,
            "updated_at": now,
        }

        await db.profiles.insert_one(profile_doc)

        # Mark user profile as complete
        await db.users.update_one(
            {"user_id": user_id},
            {"$set": {"profile_complete": True, "updated_at": now}}
        )

        return UserProfile(
            user_id=user_id,
            email=user["email"],
            role=AccountRole.TEEN,
            name=data.name,
            age=data.age,
            height=data.height,
            weight=data.weight,
            icd10_code=data.icd10_code,
            advisor_name=data.advisor_name,
            profile_complete=True,
            created_at=now,
            updated_at=now,
        )

    async def create_parent_profile(self, user_id: str, data: ParentProfileCreate) -> UserProfile:
        """Create a parent profile."""
        db = get_database()

        # Get user and verify role
        user = await db.users.find_one({"user_id": user_id})
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found"
            )

        if user.get("role") != AccountRole.PARENT.value:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="User is not a parent account"
            )

        # Check if profile already exists
        existing_profile = await db.profiles.find_one({"user_id": user_id})
        if existing_profile:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="Profile already exists. Use update endpoint."
            )

        now = datetime.utcnow()

        # Create profile document
        profile_doc = {
            "user_id": user_id,
            "role": AccountRole.PARENT.value,
            "name": data.name,
            "age": data.age,
            "height": {"value": data.height.value, "unit": data.height.unit.value} if data.height else None,
            "weight": {"value": data.weight.value, "unit": data.weight.unit.value} if data.weight else None,
            "icd10_code": None,
            "advisor_name": None,
            "created_at": now,
            "updated_at": now,
        }

        await db.profiles.insert_one(profile_doc)

        # Mark user profile as complete
        await db.users.update_one(
            {"user_id": user_id},
            {"$set": {"profile_complete": True, "updated_at": now}}
        )

        return UserProfile(
            user_id=user_id,
            email=user["email"],
            role=AccountRole.PARENT,
            name=data.name,
            age=data.age,
            height=data.height,
            weight=data.weight,
            icd10_code=None,
            advisor_name=None,
            profile_complete=True,
            created_at=now,
            updated_at=now,
        )

    async def get_profile(self, user_id: str) -> UserProfile:
        """Get user profile."""
        db = get_database()

        user = await db.users.find_one({"user_id": user_id})
        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found"
            )

        profile = await db.profiles.find_one({"user_id": user_id})
        if not profile:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Profile not found. Please complete profile setup."
            )

        # Convert height/weight to proper format
        height_data = None
        if profile.get("height"):
            height_data = HeightData(
                value=profile["height"]["value"],
                unit=profile["height"]["unit"]
            )

        weight_data = None
        if profile.get("weight"):
            weight_data = WeightData(
                value=profile["weight"]["value"],
                unit=profile["weight"]["unit"]
            )

        return UserProfile(
            user_id=user_id,
            email=user["email"],
            role=AccountRole(profile["role"]),
            name=profile["name"],
            age=profile.get("age"),
            height=height_data,
            weight=weight_data,
            icd10_code=profile.get("icd10_code"),
            advisor_name=profile.get("advisor_name"),
            profile_complete=True,
            created_at=profile["created_at"],
            updated_at=profile["updated_at"],
        )

    async def update_profile(self, user_id: str, data: ProfileUpdate) -> UserProfile:
        """Update user profile."""
        db = get_database()

        profile = await db.profiles.find_one({"user_id": user_id})
        if not profile:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Profile not found"
            )

        # Build update document
        update_fields = {"updated_at": datetime.utcnow()}

        if data.name is not None:
            update_fields["name"] = data.name
        if data.age is not None:
            if data.age < 13:
                raise HTTPException(
                    status_code=status.HTTP_400_BAD_REQUEST,
                    detail="Age must be 13 or older"
                )
            update_fields["age"] = data.age
        if data.height is not None:
            update_fields["height"] = {"value": data.height.value, "unit": data.height.unit.value}
        if data.weight is not None:
            update_fields["weight"] = {"value": data.weight.value, "unit": data.weight.unit.value}
        if data.icd10_code is not None:
            # Only allow for teen accounts
            if profile["role"] == AccountRole.TEEN.value:
                update_fields["icd10_code"] = data.icd10_code if data.icd10_code else None
        if data.advisor_name is not None:
            # Only allow for teen accounts
            if profile["role"] == AccountRole.TEEN.value:
                update_fields["advisor_name"] = data.advisor_name if data.advisor_name else None

        await db.profiles.update_one(
            {"user_id": user_id},
            {"$set": update_fields}
        )

        return await self.get_profile(user_id)

    async def delete_profile(self, user_id: str) -> bool:
        """Delete user profile (soft delete would be better for production)."""
        db = get_database()

        result = await db.profiles.delete_one({"user_id": user_id})

        if result.deleted_count == 0:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Profile not found"
            )

        # Update user profile_complete status
        await db.users.update_one(
            {"user_id": user_id},
            {"$set": {"profile_complete": False, "updated_at": datetime.utcnow()}}
        )

        return True


# Singleton service instance
profile_service = ProfileService()

File: lumie_activity_app/lib/core/services/profile_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../constants/api_constants.dart';
import 'auth_service.dart';
import '../../shared/models/user_models.dart';

/// Profile service for managing user profiles
class ProfileService {
  static final ProfileService _instance = ProfileService._internal();
  factory ProfileService() => _instance;
  ProfileService._internal();

  final AuthService _authService = AuthService();

  Map<String, String> get _headers => {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer ${_authService.token}',
      };

  /// Create a teen profile
  Future<UserProfile> createTeenProfile({
    required String name,
    required int age,
    required HeightData height,
    required WeightData weight,
    String? icd10Code,
    String? advisorName,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('${ApiConstants.baseUrl}/profile/teen'),
        headers: _headers,
        body: json.encode({
          'name': name,
          'age': age,
          'height': height.toJson(),
          'weight': weight.toJson(),
          'icd10_code': icd10Code,
          'advisor_name': advisorName,
        }),
      ).timeout(const Duration(seconds: 3));

      if (response.statusCode == 200) {
        final profile = UserProfile.fromJson(json.decode(response.body));
        await _authService.updateUserState(profileComplete: true);
        return profile;
      } else {
        final error = json.decode(response.body);
        throw Exception(error['detail'] ?? 'Failed to create profile');
      }
    } catch (e) {
      // Fallback to mock for local development
      return _mockCreateTeenProfile(
        name: name,
        age: age,
        height: height,
        weight: weight,
        icd10Code: icd10Code,
        advisorName: advisorName,
      );
    }
  }

  /// Create a parent profile
  Future<UserProfile> createParentProfile({
    required String name,
    int? age,
    HeightData? height,
    WeightData? weight,
  }) async {
    try {
      final response = await http.post(
        Uri.parse('${ApiConstants.baseUrl}/profile/parent'),
        headers: _headers,
        body: json.encode({
          'name': name,
          'age': age,
          'height': height?.toJson(),
          'weight': weight?.toJson(),
        }),
      ).timeout(const Duration(seconds: 3));

      if (response.statusCode == 200) {
        final profile = UserProfile.fromJson(json.decode(response.body));
        await _authService.updateUserState(profileComplete: true);
        return profile;
      } else {
        final error = json.decode(response.body);
        throw Exception(error['detail'] ?? 'Failed to create profile');
      }
    } catch (e) {
      // Fallback to mock for local development
      return _mockCreateParentProfile(
        name: name,
        age: age,
        height: height,
        weight: weight,
      );
    }
  }

  /// Get user profile
  Future<UserProfile> getProfile() async {
    try {
      final response = await http.get(
        Uri.parse('${ApiConstants.baseUrl}/profile'),
        headers: _headers,
      ).timeout(const Duration(seconds: 3));

      if (response.statusCode == 200) {
        return UserProfile.fromJson(json.decode(response.body));
      } else {
        final error = json.decode(response.body);
        throw Exception(error['detail'] ?? 'Failed to get profile');
      }
    } catch (e) {
      // Return empty profile if no cached profile exists
      throw Exception('Failed to get profile - no local data available');
    }
  }

  /// Update user profile
  Future<UserProfile> updateProfile({
    String? name,
    int? age,
    HeightData? height,
    WeightData? weight,
    String? icd10Code,
    String? advisorName,
  }) async {
    final body = <String, dynamic>{};
    if (name != null) body['name'] = name;
    if (age != null) body['age'] = age;
    if (height != null) body['height'] = height.toJson();
    if (weight != null) body['weight'] = weight.toJson();
    if (icd10Code != null) body['icd10_code'] = icd10Code;
    if (advisorName != null) body['advisor_name'] = advisorName;

    final response = await http.put(
      Uri.parse('${ApiConstants.baseUrl}/profile'),
      headers: _headers,
      body: json.encode(body),
    );

    if (response.statusCode == 200) {
      return UserProfile.fromJson(json.decode(response.body));
    } else {
      final error = json.decode(response.body);
      throw Exception(error['detail'] ?? 'Failed to update profile');
    }
  }

  /// Search ICD-10 codes
  Future<List<ICD10Code>> searchICD10Codes(String query, {int limit = 20}) async {
    try {
      final response = await http.get(
        Uri.parse('${ApiConstants.baseUrl}/profile/icd10/search?query=$query&limit=$limit'),
        headers: _headers,
      ).timeout(const Duration(seconds: 2));

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return (data['results'] as List)
            .map((json) => ICD10Code.fromJson(json))
            .toList();
      } else {
        throw Exception('Failed to search ICD-10 codes');
      }
    } catch (e) {
      // Return mock data if API fails (for local development)
      return _getMockICD10Results(query, limit);
    }
  }

  /// Mock ICD-10 data for local development
  List<ICD10Code> _getMockICD10Results(String query, int limit) {
    final mockData = [
      ICD10Code(
        code: 'E10',
        description: 'Type 1 diabetes mellitus',
        category: 'Endocrine, nutritional and metabolic diseases',
      ),
      ICD10Code(
        code: 'E11',
        description: 'Type 2 diabetes mellitus',
        category: 'Endocrine, nutritional and metabolic diseases',
      ),
      ICD10Code(
        code: 'J45',
        description: 'Asthma',
        category: 'Diseases of the respiratory system',
      ),
      ICD10Code(
        code: 'G40',
        description: 'Epilepsy',
        category: 'Diseases of the nervous system',
      ),
      ICD10Code(
        code: 'M79.7',
        description: 'Fibromyalgia',
        category: 'Diseases of the musculoskeletal system',
      ),
      ICD10Code(
        code: 'F84.0',
        description: 'Autism spectrum disorder',
        category: 'Mental and behavioural disorders',
      ),
      ICD10Code(
        code: 'F90',
        description: 'Attention-deficit hyperactivity disorder',
        category: 'Mental and behavioural disorders',
      ),
      ICD10Code(
        code: 'Q90',
        description: 'Down syndrome',
        category: 'Congenital malformations',
      ),
      ICD10Code(
        code: 'I50',
        description: 'Heart failure',
        category: 'Diseases of the circulatory system',
      ),
      ICD10Code(
        code: 'M05',
        description: 'Rheumatoid arthritis',
        category: 'Diseases of the musculoskeletal system',
      ),
      ICD10Code(
        code: 'K50',
        description: 'Crohn\'s disease',
        category: 'Diseases of the digestive system',
      ),
      ICD10Code(
        code: 'K51',
        description: 'Ulcerative colitis',
        category: 'Diseases of the digestive system',
      ),
      ICD10Code(
        code: 'G35',
        description: 'Multiple sclerosis',
        category: 'Diseases of the nervous system',
      ),
      ICD10Code(
        code: 'M32',
        description: 'Systemic lupus erythematosus',
        category: 'Diseases of the musculoskeletal system',
      ),
      ICD10Code(
        code: 'D50',
        description: 'Iron deficiency anemia',
        category: 'Diseases of the blood',
      ),
    ];

    final lowerQuery = query.toLowerCase();
    final filtered = mockData.where((code) =>
      code.code.toLowerCase().contains(lowerQuery) ||
      code.description.toLowerCase().contains(lowerQuery) ||
      code.category.toLowerCase().contains(lowerQuery)
    ).take(limit).toList();

    return filtered;
  }

  /// Mock teen profile creation for local development
  Future<UserProfile> _mockCreateTeenProfile({
    required String name,
    required int age,
    required HeightData height,
    required WeightData weight,
    String? icd10Code,
    String? advisorName,
  }) async {
    // Simulate network delay
    await Future.delayed(const Duration(milliseconds: 500));

    final profile = UserProfile(
      userId: _authService.currentUser!.userId,
      email: _authService.currentUser!.email,
      role: AccountRole.teen,
      name: name,
      age: age,
      height: height,
      weight: weight,
      icd10Code: icd10Code,
      advisorName: advisorName,
      profileComplete: true,
      createdAt: DateTime.now(),
      updatedAt: DateTime.now(),
    );

    await _authService.updateUserState(profileComplete: true);
    return profile;
  }

  /// Mock parent profile creation for local development
  Future<UserProfile> _mockCreateParentProfile({
    required String name,
    int? age,
    HeightData? height,
    WeightData? weight,
  }) async {
    // Simulate network delay
    await Future.delayed(const Duration(milliseconds: 500));

    final profile = UserProfile(
      userId: _authService.currentUser!.userId,
      email: _authService.currentUser!.email,
      role: AccountRole.parent,
      name: name,
      age: age,
      height: height,
      weight: weight,
      profileComplete: true,
      createdAt: DateTime.now(),
      updatedAt: DateTime.now(),
    );

    await _authService.updateUserState(profileComplete: true);
    return profile;
  }
}
